lib/remote.go

```diff
@@ -29,155 +30,191 @@ func NewRemoteMethods(inst *Instance) *RemoteMethods {
+	if r.inst.http != nil {
+		res := map[string][]dsref.VersionInfo{}
+		err := r.inst.http.Call(ctx, AEFeeds, p, &res)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
 	}
-	ctx := context.TODO()
 
-	addr, err := remote.Address(r.inst.GetConfig(), *remoteName)
+	addr, err := remote.Address(r.inst.GetConfig(), p.Remote)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	feed, err := r.inst.RemoteClient().Feeds(ctx, addr)
 	if err != nil {
-		return err
+		return nil, err
 	}
-
-	*res = feed
-	return nil
+	return feed, nil
 }
 
 // PreviewParams provides arguments to the preview method
 type PreviewParams struct {
-	RemoteName string
-	Ref        string
+	Remote string
+	Ref    string
+}
+
+// UnmarshalFromRequest implements a custom deserialization-from-HTTP request
+func (p *PreviewParams) UnmarshalFromRequest(r *http.Request) error {
+	if p == nil {
+		p = &PreviewParams{}
+	}
+
+	params := *p
+	if params.Ref == "" {
+		params.Ref = r.FormValue("refstr")
+	}
+	if params.Remote == "" {
+		params.Remote = r.FormValue("remote")
+	}
+
+	*p = params
+	return nil
 }
 
 // Preview requests a dataset preview from a remote
-func (r *RemoteMethods) Preview(p *PreviewParams, res *dataset.Dataset) error {
-	if r.inst.rpc != nil {
-		return checkRPCError(r.inst.rpc.Call("RemoteMethods.Preview", p, res))
+func (r *RemoteMethods) Preview(ctx context.Context, p *PreviewParams) (*dataset.Dataset, error) {
+	if r.inst.http != nil {
+		res := &dataset.Dataset{}
+		err := r.inst.http.Call(ctx, AEPreview, p, res)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
 	}
-	ctx := context.TODO()
 
 	ref, err := dsref.Parse(p.Ref)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	addr, err := remote.Address(r.inst.GetConfig(), p.RemoteName)
+	addr, err := remote.Address(r.inst.GetConfig(), p.Remote)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	pre, err := r.inst.RemoteClient().PreviewDatasetVersion(ctx, ref, addr)
+	res, err := r.inst.RemoteClient().PreviewDatasetVersion(ctx, ref, addr)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	*res = *pre
-	return nil
+	return res, nil
 }
 
 // PushParams encapsulates parmeters for dataset publication
 type PushParams struct {
-	Ref        string
-	RemoteName string
+	Ref    string
+	Remote string
 	// All indicates all versions of a dataset and the dataset namespace should
 	// be either published or removed
 	All bool
 }
 
-// Push posts a dataset version to a remote
-func (r *RemoteMethods) Push(p *PushParams, res *dsref.Ref) error {
-	if r.inst.rpc != nil {
-		return checkRPCError(r.inst.rpc.Call("RemoteMethods.Push", p, res))
+// UnmarshalFromRequest implements a custom deserialization-from-HTTP request
+func (p *PushParams) UnmarshalFromRequest(r *http.Request) error {
+	if p == nil {
+		p = &PushParams{}
+	}
+
+	params := *p
+	if params.Ref == "" {
+		params.Ref = r.FormValue("refstr")
```

Comment:
this seems to be the only thing standing in the way of being able to deserialize without the `UnmarshalFromRequest` override. Would love to get there in a follow-up PR